File: Data-structures/Queue/array-queue.py
# IMPLEMENTING QUEUES USING ARRAYS 

class Queue: 
    # Initialize the object 
    def __init__(self, c): 
        self.queue = [] 
        self.front = self.rear = 0 
        self.capacity = c 

    # Insert an element at the end(rear) of the queue (Enqueue)
    def insertElement(self, data): 
        # Check whether the queue is full
        if (self.capacity == self.rear): 
            print("\nQueue is full") 
        else: 
            self.queue.append(data) 
            self.rear += 1 

    # Function to delete an element from the front of the queue (Dequeue) 
    def deleteElement(self): 
        # Check whether the queue is empty 
        if (self.front == self.rear): 
            print("Queue is empty") 
        else: 
            x = self.queue.pop(0) 
            self.rear -= 1 

    # Display the queue elements 
    def displayQueue(self):
        if(self.rear == self.front): 
            print("\nQueue is empty") 
        else: 
            for i in self.queue: 
                print(i,'<-',end="")

    # Print the first element in the queue (front)
    def displayFront(self): 
        if (self.front == self.rear):
            print("\nQueue is empty") 
        
        print("\nFront element is: ", 
          self.queue[self.front])

# Driver code 
if __name__ == '__main__': 
    print("Queue in python")
    #  Make a new queue of capacity 5 
    newQueue = Queue(5) 

    # Display the queue elements 
    newQueue.displayQueue() 

    # Insert elements into the queue 
    newQueue.insertElement(10)
    newQueue.insertElement(20)
    newQueue.insertElement(30)
    newQueue.insertElement(40)
    newQueue.insertElement(50) 

    # Display the inserted elements 
    newQueue.displayQueue() 

File: Data-structures/Queue/Javascript/array-queue.js
// IMPLEMENTING QUEUES IN JAVASCRIPT USING ARRAYS 

class Queue {
    constructor() {
        this.data = []
    }

    // Enqueue method for adding elements into the queue 
    enqueue(item) {
        this.data.push(item);
        console.log(`${item} as just been added to the queue`); 
    }  

    // dequeue method for removing elements from the queue 
    dequeue() {
        if (this.data.length === 0){
            console.log("The queue is already empty")
        } else {
            let removedItem = this.data[0]; 
            this.data.pop(); 
            console.log(`${removedItem} has just been removed from the queue`); 
        }
    }

    // Method to check whether the queue is empty 
    isEmpty() {
        if (this.data.length === 0){
            console.log("The queue is empty"); 
        } else {
            console.log("The queue is not empty"); 
        }
    } 

    // Method to check the topmost element
    checkTop() {
        if (this.data.length === 0){
            console.log("Nothing here to see, the queue is empty!"); 
        } else {
            let topElement = this.data[0]; 
            console.log(`The topmost element is ${topElement}`); 
        }
    }

    // Function for printing the queue
    printQueue() {
        let queue = ""; 
        this.data.forEach( item => {
            queue += item + " "; 
        })
        console.log(queue); 
    } 


} 

let queue = new Queue(); 

// Adding items to the queue 
queue.enqueue(10)
queue.enqueue(20)
queue.enqueue(30)
queue.enqueue(40)
queue.enqueue(50)

// Checking the topmost element 
queue.checkTop(); 

// Removing item from the queue 
queue.dequeue(); 

// Print all the items in thq queue 
queue.printQueue(); 

